pkg/libvmi/cloudinit.go

```diff
@@ -86,16 +86,33 @@ func addDiskVolumeWithCloudInitConfigDrive(vmi *v1.VirtualMachineInstance, diskN
 	addDisk(vmi, newDisk(diskName, bus))
 	v := newVolume(diskName)
 	v.VolumeSource = v1.VolumeSource{CloudInitConfigDrive: &v1.CloudInitConfigDriveSource{}}
-	addVolume(vmi, v)
+	addCloudInitVolume(vmi, v)
 }
 
 func addDiskVolumeWithCloudInitNoCloud(vmi *v1.VirtualMachineInstance, diskName string, bus v1.DiskBus) {
 	addDisk(vmi, newDisk(diskName, bus))
 	v := newVolume(diskName)
 	setCloudInitNoCloud(&v, &v1.CloudInitNoCloudSource{})
-	addVolume(vmi, v)
+	addCloudInitVolume(vmi, v)
 }
 
 func setCloudInitNoCloud(volume *v1.Volume, source *v1.CloudInitNoCloudSource) {
 	volume.VolumeSource = v1.VolumeSource{CloudInitNoCloud: source}
 }
+
+func addCloudInitVolume(vmi *v1.VirtualMachineInstance, volume v1.Volume) {
+	if !cloudInitVolumeExists(vmi, volume) {
```

Comment:
I don't understand the bug that this commit tries to fix, but I don't believe that this is an acceptable semantics: you should not silently keep the former volume.